pkgs/development/compilers/gcc/common/configure-flags.nix

```
@@ -114,6 +114,14 @@ let
     ++ lib.optional (!(crossMingw && crossStageStatic))
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+    # cross-built gcc needs to be tricked as if it is built with headers
+    # available in sysroot. For that we need to pass some --with-sysroot= value:
+    #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+    # Without this flag --with-native-system-header-dir= has no effect
+    # and target libc features are not detected like stack protector support
+    ++ lib.optional (buildPlatform != hostPlatform && hostPlatform == targetPlatform)
+      "--with-sysroot=/"
```

```suggestion
      # When host!=build or build!=host, gcc ignores --native-system-header-dir
      # unless we provide --with-sysroot, which will be automatically prepended
      # to it.  Since we pass a store path for --native-system-header-dir, which
      # is always an absolute path, prepending an additional "/" has no effect.
      "--with-sysroot=/"
```

Your trick is very clever!  In fact it is so clever that I think we should explain it in a comment in the code in addition to the commit message.  I read the code first and was very confused; then I read the commit message and it all made sense.

---

I reworded the comment a bit. Does it read better now?

---

Added explicit mention of offset prefix and aded few examples of prefixed gcc uses.

---

Fantastic, thank you.